### PR TITLE
pybind/mgr/CMakeLists.txt: exclude files not used at runtime

### PR DESCRIPTION
**Pull request by a16bitsysop**
Exclude node_modules so it is not installed with make install
Exclude cypress files and .editorconfig also

Fixes: https://tracker.ceph.com/issues/50827
Signed-off-by: Duncan Bellamy <dunk@denkimushi.com>

## Checklist
- [X] References tracker ticket
- [X] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
on amd64:
```console
$ ls -lh bin/ceph-dencoder
-rwxrwxr-x. 1 kchai kchai 744M May 15 11:23 bin/ceph-dencoder
```

**Comment by a16bitsysop** [comment]
Thanks I will try removing that as well

**Comment by a16bitsysop** [comment]
There are more hidden (.) files in `usr/share/ceph/mgr/dashboard` and
`usr/share/ceph/mgr/dashboard/frontend` as well

**Comment by tchaikov** [comment]
@a16bitsysop 

> There are more hidden (.) files in `usr/share/ceph/mgr/dashboard` and
> `usr/share/ceph/mgr/dashboard/frontend` as well

```console
$ pwd
/home/kchai/ceph/src/pybind/mgr/dashboard
$ find . -type f -name '.*' -exec du -ch {} + | grep total
3.3M	total
```

@a16bitsysop i encourage you to find the major contributors in the sense of total size of the installation, not the number.

**Comment by tchaikov** [comment]
the dot files installed by various npm packages **can** be ignored when redistributing them, but i think the right way is let the authors of those packages to decide which files should be included / excluded when the npm is published. npm checks for `.gitignore`, `.npmignore` and the `file` field in `package.json`. some (versions) of some npm projects might not have a very finer grained setting in this regard, but i am afraid if we remove all dot files without carefully auditing the removed files, the resulting npm packages might fail to work. anyway, i'd defer to this to @epuertat .

**Comment by a16bitsysop** [comment]
Okay, I tried lots of things in the alpine ci and it always runs out of space on aarch64. I even patched out ceph-dencoder compiling.

Yes I wasn’t sure which files where required.

**Comment by a16bitsysop** [comment]
Will have to have another look and see what else is using the space

**Comment by tchaikov** [comment]
@a16bitsysop thanks. might want to sort by the size of artifacts in the install directory when the build is out-of-space.

File: src/pybind/mgr/CMakeLists.txt
Comment (epuertat) [inline src/pybind/mgr/CMakeLists.txt:28]: Not sure if dot files/dirs are copied by cmake install (@tchaikov can you please confirm?), but it might make sense to exclude all of them (e.g.: `.tox`, `.cache`, ...)?
```suggestion
  REGEX "\..+" EXCLUDE
  REGEX "requirements.*\.txt" EXCLUDE
  REGEX "constraints.*\.txt" EXCLUDE  
  REGEX "node_modules" EXCLUDE
  REGEX "\.editorconfig" EXCLUDE
  REGEX "cypress.*" EXCLUDE)
```
You can get inspiration from https://github.com/ceph/ceph/pull/31882/files#diff-97c00666af83ea265ed998e431e56b5c827b490bdaf6ba4e0c196697ccfe1478R24-R28: 
Following paths should also be excluded:
* `wheelhouse`
* `tests`
* Everything under `.../mgr/dashboard/frontend` except `.../dashboard/frontend/dist` (e.g.: `coverage`, `node_modules`, `cypress`, `src`...). I remember looking for a way to exclude everything except a dir and I couldn't find any :/

**Review by epuertat — COMMENTED** [review]
This is great, @a16bitsysop ! So this should supersede my 'orphan' PR #31882.

**Comment by epuertat** [comment]
> the dot files installed by various npm packages **can** be ignored when redistributing them, but i think the right way is let the authors of those packages to decide which files should be included / excluded when the npm is published. npm checks for `.gitignore`, `.npmignore` and the `file` field in `package.json`. some (versions) of some npm projects might not have a very finer grained setting in this regard, but i am afraid if we remove all dot files without carefully auditing the removed files, the resulting npm packages might fail to work. anyway, i'd defer to this to @epuertat .

@tchaikov I'm not an expert on how Webpack bundles all the different NPMs, but I think everything required should be under the `mgr/dashboard/frontend/dist` and nothing else. @alfonsomthd can you please confirm?

**Comment by a16bitsysop** [comment]
@epuertat thanks

**Comment by sebastian-philipp** [comment]
@kshtsk  + @mgfritch : Please make sure this works for you.

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Review by alfonsomthd — CHANGES_REQUESTED** [review]
@a16bitsysop Please update the titles of tracker issue & PR to reflect the goal of this cleanup.

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by a16bitsysop** [comment]
@alfonsomthd I can't change the tracker title, if I try and edit it just goes to empty text box at bottom of screen

**Comment by alfonsomthd** [comment]
> @alfonsomthd I can't change the tracker title, if I try and edit it just goes to empty text box at bottom of screen

@a16bitsysop I updated it with the same PR title.

**Comment by a16bitsysop** [comment]
Thanks

File: src/pybind/mgr/CMakeLists.txt
Comment (a16bitsysop) [inline src/pybind/mgr/CMakeLists.txt:25]: This seems to match everything on alpine:
https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/395447/raw

Comment (epuertat) [reply src/pybind/mgr/CMakeLists.txt:25]: Argh, perhaps it needs to be double escaped `\\..+`?

Comment (a16bitsysop) [reply src/pybind/mgr/CMakeLists.txt:25]: will try that

Comment (a16bitsysop) [reply src/pybind/mgr/CMakeLists.txt:25]: That is the same:
https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/395565/raw

Without the line it builds like this:
https://gitlab.alpinelinux.org/a16bitsysop/aports/-/jobs/395538/raw

Comment (a16bitsysop) [reply src/pybind/mgr/CMakeLists.txt]: I changed to more specific names instead of the one line

File: src/pybind/mgr/CMakeLists.txt
Comment (tchaikov) [inline src/pybind/mgr/CMakeLists.txt:33]: @a16bitsysop @epuertat if this addresses the out-of-space issue, i am wondering if a follow-up PR which enumerates 

- `src/pybind/mgr/*.py` (exclude `hello`)
- `src/pybind/mgr/**` which contains `__init__.py` at its root directory.
- `dist/**`

would be more explicit. what do you think?

Comment (a16bitsysop) [reply src/pybind/mgr/CMakeLists.txt:33]: Okay, do you mean all the python files are not needed either in mgr?

I found the alpine builders do have different free spaces, the aarch64 one has 

```
>>> Host system information (arch: aarch64, release: edge) <<<
- Number of Cores: 32
- Memory: 62.6868 Gb
- Free space: 33.2G
```

I am going to try and build locally as well

Comment (tchaikov) [reply src/pybind/mgr/CMakeLists.txt:33]: no, i want to create a PR to use the "include" list instead of "exclude" list, once yours is merged. so that only the python files and the files under `dist` are installed.

Comment (a16bitsysop) [reply src/pybind/mgr/CMakeLists.txt:33]: Ok, yes that makes sense and more future proof.

Comment (epuertat) [reply src/pybind/mgr/CMakeLists.txt:33]: > no, i want to create a PR to use the "include" list instead of "exclude" list, once yours is merged. so that only the python files and the files under dist are installed.

@tchaikov that was my initial idea here, but I later discarded  due to the risk of inadvertently missing files, since that case wouldn't be detected by regular Pull Request CI, but at the later Teuthology package-based integration testing.

In any case, your list makes sense. We would also need to include `src/pybind/mgr/dashboard/frontend/package.json`, which is used to get the default locale for the dashboard i18n.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-21_12:49:52-rados-wip-kefu-testing-2021-05-21-1851-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-22_15:42:09-rados-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

failure tracked by 

- https://tracker.ceph.com/issues/50243
- https://tracker.ceph.com/issues/50151

**Comment by tchaikov** [comment]
jenkins test dashboard

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-21_12:49:52-rados-wip-kefu-testing-2021-05-21-1851-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-22_15:42:09-rados-wip-kefu-testing-2021-05-22-2207-distro-basic-smithi/

failure tracked by 

- https://tracker.ceph.com/issues/50243
- https://tracker.ceph.com/issues/50151

**Comment by epuertat** [comment]
Dashboard tests:
* Passed: 18
* Failed: 20
* Skipped: 61
* Total: 99

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
[Dashboard tests](https://dashboard.cypress.io/projects/k7ab29/runs/3225/overview):
* Passed: 85
* Failed: 4
* Skipped: 10
* Total: 99

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard

**Review by epuertat — APPROVED** [review]
LGTM! Thanks a lot for this clean-up, @a16bitsysop !

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
> @a16bitsysop Please update the titles of tracker issue & PR to reflect the goal of this cleanup.

@alfonsomthd could you please take another look? i think if a PR only contains a single commit, it might be advisable to reuse the title of that commit for the title of the PR. and in general, the title of a commit should summarize the change using an imperative mood. 

@a16bitsysop i think what we can improve is the body of the commit message. i.e., to explain that the intent of this change is to reduce the total size of installed files by dropping unused files.

**Review by alfonsomthd — APPROVED** [review]

**Comment by a16bitsysop** [comment]
@tchaikov  I can't update the title in the tracker, I don't have enough permission to do it.

The other thing I did in alpine was to set CFLAGS and CPPFLAGS with:

```
export CFLAGS="$CFLAGS -s"
export CXXFLAGS="$CXXFLAGS -s"
```

as:
```
//Default BUILD_TYPE is Debug, other options are: RelWithDebInfo,
// Release, and MinSizeRel.
```

So it was building everything with debug info, and filling up the free disc space.  I am going to try and set BUILD_TYPE to MinSizeRel on alpine, as BUILD_TYPE is supposed to be none on alpine.

**Comment by tchaikov** [comment]
@a16bitsysop i think the title of the tracker ticket is good enough from my point of view. it reads like a solution though.

> So it was building everything with debug info, and filling up the free disc space. I am going to try and set BUILD_TYPE to MinSizeRel on alpine, as BUILD_TYPE is supposed to be none on alpine.

do you think we should leave `CMAKE_BUILD_TYPE` empty when building from a dist tarball? if it's an encouraged practice, i can help on this.

**Comment by a16bitsysop** [comment]
@tchaikov yes that would help if it could be empty when building from a dist tarball thanks, it is the encouraged practice for alpine so they can use their own cflags.

Looking at there defaults:
https://git.alpinelinux.org/abuild/tree/abuild.conf

There is no default BUILD_TYPE for cmake, which I guess is why they want BUILD_TYPE to be none so there cflags are used?

It looks like debug symbols are added for alpine if a -dbg sub package exists.

A bit more info here:
https://wiki.alpinelinux.org/wiki/Release_Notes_for_Alpine_3.12.0#The_.22plain.22_buildtype_for_Meson_and_.22None.22_buildtype_for_CMake_are_used_now

**Comment by tchaikov** [comment]
created #42129 . i foresee this change would be controversial though.

**Comment by a16bitsysop** [comment]
Thanks a lot, hope it is accepted

File: src/pybind/mgr/CMakeLists.txt
Comment (tchaikov) [reply src/pybind/mgr/CMakeLists.txt:33]: #42255

**Pull request merged**
